source/simulator/core.cc

```diff
@@ -506,9 +495,18 @@ namespace aspect
         //use it, we can run into problems with geometry_model->depth().
         AssertThrow ( parameters.pressure_normalization == "no",
                       ExcMessage("The free surface scheme can only be used with no pressure normalization") );
-        free_surface.reset( new FreeSurfaceHandler( *this, prm ) );
+
+        //Allocate the FreeSurfaceHandler object
+        free_surface.reset( new FreeSurfaceHandler<dim>( *this, prm ) );
       }
 
+    //The mapping is given by a degree four MappingQ for the case
+    //of a curved mesh, and a degree one MappingQ for a rectangular mesh.
+    //If a free surface is enabled, this is swapped out for a MappingQ1Eulerian,
+    //which allows for mesh deformation.
+    mapping.reset( new MappingQ<dim>( geometry_model->has_curved_elements() ? 4 : 1,
+                                      geometry_model->has_curved_elements() ? true : false ) );
```

Comment:
You are right. I originally had this as a branch statement rather than the current swapping behavior. This can be back in the initializer list. The moment when the `FreeSurfaceHandler` swaps in the `MappingQEulerian` is kind of tricky: it needs to happen after the DoFs have been set up, since I need to give it a valid vector, but it needs to happen before various boundary conditions are calculated, since the mapping needs to be valid.